jlm/rvsdg/tracker.cpp

```diff
@@ -116,7 +127,7 @@ class TrackerDepthState
   size_t count_;
   size_t top_depth_;
   size_t bottom_depth_;
-  std::unordered_map<size_t, std::unordered_set<TrackerNodeState *>> nodestates_;
+  std::unordered_map<size_t, std::set<TrackerNodeState *, cmp>> nodestates_;
```

Comment:
no that does not work -- comparator for unordered_map must be an equality comparison, _not_ less than.

The problem here is that the hash and comparison is based on addresses, not ids. If you want to make it deterministic, you need to either:

- use std::map, and then used less than
- rework the unordered_map to be based solely on  ids

The critical operations are peek_top and peek_bottom. There is nothing that you can do by introducing node ids or anything to make these deterministic (even _with_ node id, you are still iterating a hash, and it does _not_ guarantee determinism across executions).

Only a change of data structure will help here.

Reply:
oops yes I got confused with the nesting, that should be fine then

I would however note that determinism here could have been achieved entirely without node id, just using some double data structure (node held in list in queue order, association with list iterator via hash table)